src/computability/GNFA.lean

```diff
@@ -0,0 +1,675 @@
+      revert x_matches,
+      rw ← xs.reverse_reverse,
+      induction xs.reverse,
+      case nil
+      { assume matches,
+        rw list.reverse_nil,
+        unfold list.join,
+        rw list.append_nil,
+        refine trace.step _ z_matches rfl,
+        rw rip_trace_correct,
+        exact t, },
+      case cons : x xs ih
+      { assume matches,
+        rw [list.reverse_cons, list.join_append],
+        unfold list.join,
+        rw [list.append_nil, ← list.append_assoc],
+        simp only [list.mem_reverse, list.mem_cons_iff] at matches,
+        refine trace.step _ (matches x (or.inl rfl)) rfl,
+        apply ih,
+        assume x mem,
+        rw list.mem_reverse at mem,
+        exact matches x (or.inr mem), }, }, },
+  { assume t,
+    cases t,
+    case start : x matches { exact accepts.start (or.inl matches), },
+    case step : x y z q t matches eq
+    { rw eq, clear eq x,
+      cases M.rip_trace_aux t,
+      case inl
+      { rcases h with ⟨q, eq, t⟩,
+        rw eq at matches, clear eq,
+        exact accepts.step _ t (or.inl matches) rfl, },
+      rcases h with ⟨eq, h⟩,
+      rw eq at *, clear eq,
+      rcases h with ⟨y, w, xs, p, t, w_matches, x_matches, eq⟩,
+      rw eq, clear eq,
+      cases p,
+      case none
+      { rw [option.map_none', option.get_or_else_none] at t,
+        rw t, clear t y,
+        refine accepts.start _,
+        right,
+        refine ⟨_, _, _, matches, rfl⟩,
+        refine ⟨_, _, w_matches, _, rfl⟩,
+        exact ⟨xs, rfl, x_matches⟩, },
+      case some
+      { rw [option.map_some', option.get_or_else_some] at t,
+        rw [list.append_assoc, list.append_assoc],
+        refine accepts.step _ t _ rfl,
+        right,
+        rw ← list.append_assoc,
+        refine ⟨_, _, _, matches, rfl⟩,
+        refine ⟨_, _, w_matches, _, rfl⟩,
+        exact ⟨xs, rfl, x_matches⟩, }, }, },
+end
+
+/--
+Convert a GNA to a regular expression by repeatedly removing internal states. When there are no
+internal states left, there will only be one transition, from the starting state to the accepting
+state. Its regular expression will accept the same language as the original GNFA.
+-/
+def to_regular_expression : Π {n}, GNFA α n → regular_expression α
+| 0 M := M.step none none
+| (nat.succ n) M := M.rip.to_regular_expression
+
+theorem to_regular_expression_correct (M : GNFA α n) :
+  M.accepts = M.to_regular_expression.matches :=
+begin
+  induction n,
+  case zero
+  { ext,
+    split,
+    { assume hx,
+      cases hx,
+      case start : x matches { exact matches, },
+      case step : x y z q t matches eq { exact fin.elim0 q, }, },
+    { assume matches,
+      exact accepts.start matches, }, },
+  case succ : n ih
+  { rw ← M.rip_correct,
+    rw ih M.rip,
+    refl, },
+end
+
+end GNFA
+
+namespace NFA
+
+/--
+Given an equivalence between `σ` and `τ`, convert an NFA with state type `σ` into the corresponding
+NFA with state type `τ`.
+
+TODO: possibly move to computability/NFA
```

Comment:
Since this is your first PR and it's rather large; I recommend you split this definition and the lemma below into a second PR (adding it to the NFA file), and then come back to this one once that's merged.